// spiritjs/lib/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.version = exports.load = exports.create = exports.groups = exports.setup = exports.config = undefined;

var _group = require('./group');

Object.keys(_group).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _group[key];
    }
  });
});

var _parser = require('./data/parser');

Object.defineProperty(exports, 'create', {
  enumerable: true,
  get: function get() {
    return _parser.create;
  }
});
Object.defineProperty(exports, 'load', {
  enumerable: true,
  get: function get() {
    return _parser.load;
  }
});

var _utils = require('./utils');

var _config2 = require('./config/config');

var _config3 = _interopRequireDefault(_config2);

var _setup2 = require('./config/setup');

var _setup3 = _interopRequireDefault(_setup2);

var _registry = require('./registry/registry');

var _registry2 = _interopRequireDefault(_registry);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var version = '1.3.0';

exports.config = _config3.default;
exports.setup = _setup3.default;
exports.groups = _registry2.default;
exports.version = version;


if (_utils.debug) {
  console.warn('You are running the development build of Spirit.');
}
